packages/core/src/storage/vectorStore/utils.ts

```diff
@@ -0,0 +1,59 @@
+import { BaseNode, Metadata, ObjectType, jsonToNode } from "../../Node";
+
+const DEFAULT_TEXT_KEY = "text";
+
+export function validateIsFlat(obj: { [key: string]: any }): void {
+  for (let key in obj) {
+    if (typeof obj[key] === "object" && obj[key] !== null) {
+      throw new Error(`Value for metadata ${key} must not be another object`);
+    }
+  }
+}
+
+export function nodeToMetadata(
+  node: BaseNode,
+  removeText: boolean = false,
+  textField: string = DEFAULT_TEXT_KEY,
+  flatMetadata: boolean = false,
+): Metadata {
+  const nodeObj = node.toJSON();
+  const metadata = node.metadata;
+
+  if (flatMetadata) {
+    validateIsFlat(node.metadata);
+  }
+
+  if (removeText) {
+    nodeObj[textField] = "";
+  }
+
+  nodeObj["embedding"] = null;
+
+  metadata["_node_content"] = JSON.stringify(nodeObj);
+  metadata["_node_type"] = node.constructor.name.replace("_", ""); // remove leading underscore to be compatible with Python
+
+  metadata["document_id"] = node.sourceNode?.nodeId || "None";
+  metadata["doc_id"] = node.sourceNode?.nodeId || "None";
+  metadata["ref_doc_id"] = node.sourceNode?.nodeId || "None";
+
+  return metadata;
+}
+
+export function metadataDictToNode(metadata: Metadata): BaseNode {
+  const nodeContent = metadata["_node_content"];
+  if (!nodeContent) {
+    throw new Error("Node content not found in metadata.");
+  }
+  const nodeObj = JSON.parse(nodeContent);
```

Comment:
@yisding it works to reconstruct objects from Python 0.9, but there is an issue that the node attributes in Python are snake_case and in TS they are camelCase, so a reconstructed `TextNode` in TS (coming from Python) looks like this:

```
_TextNode {
      id_: '95d3f69c-4197-41ca-9f87-b78d9892867f',
      metadata: {},
      excludedEmbedMetadataKeys: [],
      excludedLlmMetadataKeys: [],
      relationships: [Object],
      hash: '4JN1LIB89VS6Hq1tZbtoYrJDCN0uLyHs68Et8K4B/ZA=',
      embedding: null,
      excluded_embed_metadata_keys: [],
      excluded_llm_metadata_keys: [],
      text: "@bradleymeck @aredridel @nebrius @addaleax All fair points. We couldn't figure out a way to get around HTTP so we think bundles will still be used as a perf optimization, but getting away without a build step at development time is a worthwhile win, I think.",
      start_char_idx: null,
      end_char_idx: null,
      text_template: '{metadata_str}\n\n{content}',
      metadata_template: '{key}: {value}',
      metadata_seperator: '\n',
      class_name: 'TextNode',
      metadataSeparator: '\n'
    }
    ```
    I think we should either use snake_case in TS too or add a mapping

Reply:
Let's create a TS class to track and convert for now. The interface on the python side may also not be entirely stable yet, so let's create one that says TextNodev09 etc.